stan/math/rev/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,198 @@
+#ifndef STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/rev/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/inverse_spd.hpp>
+#include <stan/math/rev/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/mdivide_right_spd.hpp>
+#include <stan/math/rev/fun/inverse.hpp>
+
+namespace stan {
+namespace math {
+
+/*
+ * Reverse mode specialization of calculating the generalized inverse of a
+ * matrix.
+ *
+ * @param G specified matrix
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ *
+ * @note Reverse mode differentiation algorithm reference:
+ *
+ * <ul><li> Golub, G.H. and Pereyra, V. The Differentiation of Pseudo-Inverses
+ * and Nonlinear Least Squares Problems Whose Variables Separate. <i>SIAM
+ * Journal on Numerical Analysis</i>, Vol. 10, No. 2 (Apr., 1973), pp.
+ * 413-432</li></ul>
+ *
+ * Equation 4.12 in the paper
+ *
+ *  See also
+ *  http://mathoverflow.net/questions/25778/analytical-formula-for-numerical-derivative-of-the-matrix-pseudo-inverse
+ *
+ *  Implementation is based on
+ *   Title: Uncertainties Python Package
+ *    Author: Eric O. LEBIGOT
+ *    Date: 2020
+ *    Availability:
+ * https://github.com/lebigot/uncertainties/blob/master/uncertainties/unumpy/core.py
+ */
+template <typename VarMat, require_rev_matrix_t<VarMat>* = nullptr>
+inline auto generalized_inverse(const VarMat& G) {
+  using value_t = value_type_t<VarMat>;
+  using ret_type = promote_var_matrix_t<VarMat, VarMat>;
+
+  if (G.size() == 0)
+    return ret_type(G);
+
+  if (G.rows() == G.cols())
+    return ret_type(inverse(G));
+
+  if (G.rows() < G.cols()) {
+    arena_t<VarMat> G_arena(G);
+    auto A_spd = tcrossprod(G_arena.val_op());
+    arena_t<VarMat> inv_G(
+        mdivide_left_spd(A_spd.val_op(), G_arena.val_op()).transpose());
```

Comment:
1. You don't need to use `.val()/.val_op()/etc` on matrices of non-var types. When a matrices type is an `Eigen::Matrix<var, R, C>` we use `.val()`, and `.adj()`to pull out the values or adjoints of the underlying `var`s. It effectively generates an `Eigen::Matrix<double, R, C>` for each of the values and adjoints. But when you multiply something like `A.val() * B.adj()` your result is also going to be `Eigen::Matrix<double, R, C>`. So at that point you don't need to then use `.val()` etc on that result matrix.

2. In general it's more performant to call Eigen methods directly instead of using stan's functions. So for `tcrossprod()` you can directly do `G_arena.val_op() * G_arena.val_op().transpose()`. Similarly for `mdivide_left_spd` you can put everything together like
 ```suggestion
    arena_t<ret_type> inv_G((G_arena.val_op() * G_arena.val_op().transpose())
                                .llt()
                                .solve(G_arena.val_op())
                                .transpose());
```